src/NServiceBus.Core/Unicast/DisposableEndpointInstance.cs

```diff
@@ -0,0 +1,56 @@
+#if NETCOREAPP
+namespace NServiceBus
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Contains extension methods to create a <see cref="IDisposableEndpointInstance"/>.
+    /// </summary>
+    public static class DisposableEndpointExtensions
+    {
+        /// <summary>
+        /// Use this method to turn a starting endpoint instance into a <see cref="IDisposableEndpointInstance"/> that will stop the endpoint automatically on dispose.
+        /// If the endpoint is stopped during disposal, it will use a completed <see cref="CancellationToken"/>, indicating that it expects immediate shutdown.
+        /// </summary>
+#pragma warning disable PS0018 // A task-returning method should have a CancellationToken parameter unless it has a parameter implementing ICancellableContext
+        public static async Task<IDisposableEndpointInstance> AsDisposable(this Task<IEndpointInstance> startTask)
+#pragma warning restore PS0018 // A task-returning method should have a CancellationToken parameter unless it has a parameter implementing ICancellableContext
+        {
+            var runningInstance = await startTask.ConfigureAwait(false);
+            return new DisposableEndpointInstance(runningInstance);
+        }
+    }
+
+    /// <summary>
+    /// an <see cref="IEndpointInstance"/> that implements <see cref="IAsyncDisposable"/>. The endpoint performs a graceful shutdown when being disposed.
+    /// </summary>
+    public interface IDisposableEndpointInstance : IEndpointInstance, IAsyncDisposable
+    {
+    }
+
+    class DisposableEndpointInstance : IDisposableEndpointInstance
+    {
+        IEndpointInstance endpointInstance;
+
+        public DisposableEndpointInstance(IEndpointInstance endpointInstance) => this.endpointInstance = endpointInstance;
+
+        ValueTask IAsyncDisposable.DisposeAsync() => new ValueTask(endpointInstance.Stop(new CancellationToken(true)));
```

Comment:
that should be CancellationToken.None I guess, since we want to default to a graceful shutdown. The `AsDisposable` API could allow user-control of the shutdown behavior when disposing though.